docs/protocol.md

```diff
@@ -0,0 +1,282 @@
+
+Code references used:
+- `rai/node/common.cpp`
+  - `rai::message`
+  - `rai::message::read_header()`
+  - `rai::message::block_type_set()`
+- `rai/node/common.hpp`
+  - `enum class message_type`
+- `rai/lib/blocks.hpp`
+  - `enum class block_type`
+
+## Values
+
+The following different types of values are used in the protocol:
+
+  * Account
+    - The 256-bit Ed25519 public key of an account
+    - Encoded as a sequence of 32 bytes
+    - Note that the well-known `xrb_...` notation for an address/representative
+      is a different encoding of this value
+  * Representative
+    - Same as Account, but in a different role
+  * Block hash
+    - The 256-bit hash of a block  
+    - The hash is computed from a set of "hashables": fields in a block that are hashed
+      with the Blake2b hashing function
+    - Unless noted other the hashables for a block are the block-specific fields
+      listed in the table with each block description. They are to be hashed in the
+      order they are listed.
+    - The block hash is used (a.o.) to make references among blocks
+    - Encoded as a sequence of 32 bytes
+  * Balance / Amount
+    - A 128-bit unsigned integer
+    - This number has unit "raw", where 1 XRB = 1 Mxrb = 10<sup>30</sup> raw.
+      See [here](https://github.com/nanocurrency/raiblocks/wiki/Distribution,-Mining-and-Units)
+      for more information.
+    - Encoded as a sequence of 16 bytes, *big-endian* order
+  * Signature
+      - A 512-bit Ed25519 signature
+      - Encoded as a sequence of 64 bytes
+  * Work
+      - A 64-bit Proof-of-Work value
+      - Encoded as a *big-endian* unsigned integer
+
+The Ed25519 implementation in Nano does not use the standard
+SHA-512 hashing algorithm, but Blake2b. See [here](https://github.com/nanocurrency/raiblocks/wiki/Design-features)
+for more information.
+
+## UDP packets
+
+There are currently 4 types of messages sent in UDP packets:
+
+  | Type       | Description                                  |
+  | ---------- | -------------------------------------------- |
+  | Keepalive  | Status notification of peers being alive     |
+  | Publish    | Publication of a (new?) block on the network |
+  | ConfirmReq | Request to vote on a block?                  |
+  | ConfirmAck | Vote by a representative on a block         |
+
+There are other types of messages denoted by `message_type`, but these are
+only used in the "bootstrapping" procedure when a node needs to synchronize
+itself with all the blocks in the distributed ledger (in a bulk fashion).
+Bootstrapping is done over TCP connections, not with UDP packets.
+
+All UDP messages have a fixed-size header (see `rai::message::read_header()` in `rai/node/common.cpp`):
+
+  | Field         | Size (bytes)  | Description |
+  | ------------- | ------------- | ----------- |
+  | Magic         | 2             | Magic number to indicate this is a Nano packet |
```

Comment:
Technically the magic number is only one byte, 'R' in ASCII, and the second byte is another magic number ('A', 'B', 'C' in ASCII) denoting the network (live, beta, test respectively)

Reply:
Yeah that's described a bit below.

Reply:
Right just saw that hah.